src/goto-symex/symex_dereference.cpp

```diff
@@ -271,10 +328,29 @@ void goto_symext::dereference_rec(exprt &expr, statet &state, bool write)
       dereference.dereference(tmp1, symex_config.show_points_to_sets);
     // std::cout << "**** " << format(tmp2) << '\n';
 
-    expr.swap(tmp2);
 
     // this may yield a new auto-object
-    trigger_auto_object(expr, state);
+    trigger_auto_object(tmp2, state);
+
+    // If the dereference result is not a complicated expression
+    // (i.e. of the form
+    //   [let p = <expr> in ]
+    //   (p == &something ? something : ...))
+    // we should just return it unchanged.
+    // also if we are on the lhs of an assignment we should also not attempt to
+    // go to the cache we cannot do this for quantified expressions because this
+    // would result in us referencing quantifier variables outside the scope of
+    // the quantifier.
```

Comment:
Thank you for the detailed comment, but may I ask for some commas and/or full stops? It's a bit hard to read as is.

Reply:
Fixed.